### PR TITLE
Repair fix coords

### PR DESCRIPTION
**Pull request by bertsky**
This attempts to fix problems caused by invalid polygons from `ocrd-segment-repair` (both in `sanitize` and `plausibilize` mode).

This taught me another lesson about what can go wrong with Shapely / numpy / PAGE interaction. To sum up:

1. Ensuring valid polygons on the input side (e.g. from OpenCV) is always necessary. The only generic way I can think of is to feed them through `simplify` with ever increasing tolerance until valid. **EDIT2** The problem is that the result of the algorithm implemented in Shapely/GEOS depends on the starting point it picked. In pathological cases, _no simplification whatsoever_ can be achieved. (The only thing that then helps is re-ordering...)
2. Operations like `union` or `intersection` can create collections of shapes. **EDIT** There are actually 2 cases here:
   1. homogeneous ([MultiPolygon](https://shapely.readthedocs.io/en/stable/manual.html#MultiPolygon)) – a discontiguous collection of Polygon – in which case one needs the convex hull.
   2. heterogeneous ([GeometryCollection](https://shapely.readthedocs.io/en/stable/manual.html#collections)) – a collection of Polygon with Point or LineString – in which case one needs to filter out those shapes which have no intrinsic area (and then check again for the other cases)
3. Operations like `union` or `intersection` can create non-integer points, which when rounded for PAGE serialization can become invalid paths. Unfortunately, Shapely _always_ calculates in floating point internally. So all we can do is rounding and then ensuring validity (as in 1).

Related:
- https://github.com/cisocrgroup/ocrd_cis/issues/67
- https://github.com/cisocrgroup/ocrd_cis/issues/62
- https://github.com/OCR-D/ocrd_tesserocr/issues/149
- https://github.com/OCR-D/ocrd_tesserocr/issues/151

### CODE REVIEW

**Review by stweil — APPROVED** [review]
I used this PR in my test for https://github.com/OCR-D/ocrd_tesserocr/issues/149, and it seems to work.

**Comment by stweil** [comment]
@bertsky, you are faster with pushing new commits to existing pull requests than I am able to test them.

I suggest to wait a little with new commits until the existing pull request was merged and make a new one then (unless there is an urgent need for the new commit of course).

**Comment by stweil** [comment]
@kba (or whoever has merge rights), can this PR be merged? I'd like to have a working `ocrd_all` again, and this seems to be one required component.

**Comment by bertsky** [comment]
> @bertsky, you are faster with pushing new commits to existing pull requests than I am able to test them.
> 
> I suggest to wait a little with new commits until the existing pull request was merged and make a new one then (unless there is an urgent need for the new commit of course).

Sorry, there are so many related issues at the moment and the shared problem is always our Shapely code. Also, I don't have the time right now to go forward and backward on each one of them. So I felt like rushing it – this time.

> @kba (or whoever has merge rights), can this PR be merged? I'd like to have a working `ocrd_all` again, and this seems to be one required component.

Right now I am still testing myself. Also, after delegating to the PAGE validator in core, I found another bug there...

File: ocrd_segment/repair.py
Comment (kba) [inline ocrd_segment/repair.py:350]: ocrd_utils.image?

File: requirements.txt
Comment (kba) [inline requirements.txt]: 2.16.2 has the fix for empty reading order

**Review by kba — APPROVED** [review]

File: ocrd_segment/repair.py
Comment (bertsky) [reply ocrd_segment/repair.py:350]: You mean to cover all of the above in OCR-D/core#492 or OCR-D/core#577? I am not so sure – remember we actually wanted to fix this at the level of the PAGE model (whenever adding elements), throw a specialized exception in each unique case, collect page-wise exceptions in Processor.process_pages, turning them into error messages, and opt-in to seeing the stack traces with `--debug`.

File: requirements.txt
Comment (bertsky) [reply requirements.txt]: thx, update is coming...

File: ocrd_segment/repair.py
Comment (kba) [reply ocrd_segment/repair.py:350]: > remember we actually wanted to fix this at the level of the PAGE model

I am aware of and working on that as time permits. But since you have written these pretty generic utility methods like `make_valid(polygon)`, I still think it would be useful to offer them to all the OCR-D projects. Just an offer, proceed as you think is best.

Comment (bertsky) [reply ocrd_segment/repair.py:350]: A function like `make_valid` is too low-level IMHO. So far we don't pass/return Shapely objects anywhere. As for `polygon_for_parent`, I don't think it would be a good investment of resources to rewrite other processors with that stopgap, too. We should instead concentrate on the encapsulated, automatic solution in core – as time permits of course.

**Pull request merged**
